1664: management command to create Order from enrollment

collinpreston opened the pull request:
#### Pre-Flight checklist

- [x] Testing
  - [x] Changes have been manually tested

#### What are the relevant tickets?
https://github.com/mitodl/mitxonline/issues/1664

#### What's this PR do?
A recent PR ([ref](https://github.com/mitodl/mitxonline/pull/1644)) now creates an Order object when a user enrolls into a course.  The Order object's information is synced with HubSpot which allows Marketing team members to view user enrollment information.  **This PR** intends to add a management command that will create an Order for each of the existing audit enrollments.

#### How should this be manually tested?

1. Setup a Course, CourseRun, Product, User, and the CMS page for the course.
2. Configure the required HubSpot environment variables in your `.env` file.
3. Enroll your user into the Audit track of the course by clicking the "Enroll now" button but not purchasing the course.
4. Through Django Admin, delete the Order associated with your enrollment: http://mitxonline.odl.local:8013/admin/ecommerce/order/
5. In a terminal and from the root of the project, run `docker-compose run web ./manage.py create_order_from_course_run_enrollments`
6. Verify that an Order is created for your enrollment.
7. Run step 5 again and confirm that only a single Order exists for your enrollment.
8. Verify that the Order details are synced with HubSpot.
9. Upgrade your course enrollment to the Verified track by paying for the course.
10. Run step 5 again and confirm that only a single fulfilled Order exists for your enrollment: http://mitxonline.odl.local:8013/admin/ecommerce/fulfilledorder/


#### Any background context you want to provide?
Users that have a "verified" enrollment, will already have an Order associated with their enrollment since, prior to the recent change of Orders being created during enrollment, Orders were historically created at the time of initiating the checkout process.

A second management command that already exists in MITx Online, `sync_db_with_hubspot`, can be used to sync all of the Orders with HubSpot.

mbertrand reviewed (commented):
One thing I noticed though is that `PendingOrder.create_from_product` is called for every audited enrollment, even if a pending order already exists. Although it doesn't create a dupe order (just gets the existing one), it still triggers a hubspot sync.  Probably not a big deal, just maybe a bit inefficient.  However, assuming there are thousands of missing orders on production, I wonder if there's a chance we might go over the hubspot rate limit.  That can be fixed later though by calling sync_db_to_hubspot to batch create/update them.

mbertrand commented on hubspot_sync/management/commands/create_order_from_course_run_enrollments.py:
```diff
@@ -0,0 +1,47 @@
+                    lines__purchased_content_type_id=product_content_type,
+                    lines__product_version=product_version,
+                )
+                if not existing_fulfilled_order:
```
Maybe query for a PendingOrder too and then:

```python
if not existing_fulfilled_order and not existing_pending_order:
```

?

mbertrand replied:
Or are there just not that many and this will probably just be run once anyway?

collinpreston commented:
> One thing I noticed though is that `PendingOrder.create_from_product` is called for every audited enrollment, even if a pending order already exists. Although it doesn't create a dupe order (just gets the existing one), it still triggers a hubspot sync. Probably not a big deal, just maybe a bit inefficient. However, assuming there are thousands of missing orders on production, I wonder if there's a chance we might go over the hubspot rate limit. That can be fixed later though by calling sync_db_to_hubspot to batch create/update them.

@mbertrand Yeah, that's a good point.  Do you think there is a way to avoid that?  Right now it's triggering the HubSpot sync when the Order is saved.

mbertrand commented:
> Do you think there is a way to avoid that? 

Too avoid the hubspot syncing, maybe `Order.save()` could take an optional_kwarg `hubspot_sync: bool = True` also add that kwarg to `PendingOrder._get_or_create` to pass it on from there, but with `hubspot_sync=True`

collinpreston commented:
@mbertrand Okay, so your comment had me recognize that `Order.save()` was being called twice within `PendingOrder._get_or_create()`, which would result in `sync_deal_with_hubspot()` being called twice.  After reading one of Nathan's recent discussion posts, I also began to think that calling `sync_deal_with_hubspot()` within `Order.save()` was restricting developers (as we ran into with my original design for this PR) but also stuffing more non-related functionality into `Order.save()` which should really remain pretty basic.  

So I removed `sync_deal_with_hubspot()` from within `Order.save()` and added calls to `sync_deal_with_hubspot()` wherever Orders were created or updated in more outer layer/view level methods.  While this requires more responsibility on the developer to ensure that `sync_deal_with_hubspot()` is called when Orders are created or updated, it also keeps much of our business logic in the controller level of our apps.  

This also avoids calling `sync_deal_with_hubspot()` for each Order created when running the create_order_from_course_run_enrollments command.

When testing this, I did notice that running `sync_db_to_hubspot --orders create` is throwing an error: `VALIDATION_ERROR`.  However that error is also occurring with the main branch of MITx Online.

mbertrand commented:
Sounds like a good approach to me!  

For your local instance, try running `sync_hubspot_ids_to_db` first before running `sync_db_to_hubspot`, also make sure that `MITOL_HUBSPOT_API_ID_PREFIX` in your .env file is not the same value as on RC

pdpinch commented:
How far back in time is this going to go?

collinpreston commented:
> How far back in time is this going to go?

All the way back to the first enrollment.
